### PR TITLE
Massively speed up the query returned by TI.filter_for_tis

### PR DESCRIPTION
**Pull request by ashb**
The previous query generated SQL like this:

```
WHERE (task_id = ? AND dag_id = ? AND execution_date = ?) OR (task_id = ? AND dag_id = ? AND execution_date = ?)
```

Which is fine for one or maybe even 100 TIs, but when testing DAGs at
extreme size (over 21k tasks!) this query was taking for ever (162s on
Postgres, 172s on MySQL 5.7)

By changing this query to this

```
WHERE task_id IN (?,?) AND dag_id = ? AND execution_date = ?
```

the time is reduced to 1s! (1.03s on Postgres, 1.19s on MySQL)

Even on 100 tis the reduction is large, but the overall time is not
significant (0.01451s -> 0.00626s on Postgres).

Times included SQLA query construction time (but not time for calling
filter_for_tis. So a like-for-like comparison), not just DB query time:

```python
ipdb> start_filter_20k = time.monotonic(); result_filter_20k = session.query(TI).filter(tis_filter).all(); end_filter_20k = time.monotonic()
ipdb> end_filter_20k - start_filter_20k
172.30647455298458
ipdb> in_filter = TI.dag_id == self.dag_id, TI.execution_date == self.execution_date, TI.task_id.in_([o.task_id for o in old_states.keys()]);
ipdb> start_20k_custom = time.monotonic(); result_custom_20k = session.query(TI).filter(in_filter).all(); end_20k_custom = time.monotonic()
ipdb> end_20k_custom - start_20k_custom
1.1882996069907676
```

I have also removed the check that was ensuring everything was of the
same type (all TaskInstance or all TaskInstanceKey) as it felt needless
- both types have the three required fields, so the "duck-typing"
  approach at runtime (crash if doesn't have the required property)+mypy
  checks felt Good Enough.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Review by potiuk — APPROVED** [review]
WHOA!

**Comment by potiuk** [comment]
Should we try to cherry-pick it to 1.10.13. Sounds like easy thing to do @ashb  ?

**Review by kaxil — APPROVED** [review]

**Comment by ashb** [comment]
137 on test mysql. Will retry and see if it was just unlucky. (I tested locally to get numbers, and it runs fine.)

**Comment by potiuk** [comment]
I think it was unluckly. Other MySQL Core tests worked, and 137 happens occasionally

**Comment by ashb** [comment]
k some real failures now. will look

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -1811,20 +1811,34 @@ def filter_for_tis(
...
+        first_task_id = first.task_id
+        # Common path optimisations: when all TIs are for the same dag_id and execution_date, or same dag_id
+        # and task_id -- this can be over 150x for huge numbers of TIs (20k+)
+        if all(t.dag_id == dag_id and t.execution_date == execution_date for t in tis):
+            return and_(
+                TaskInstance.dag_id == dag_id,
+                TaskInstance.execution_date == execution_date,
+                TaskInstance.task_id.in_(t.task_id for t in tis),
+            )
+        if all(t.dag_id == dag_id and t.task_id == first_task_id for t in tis):
```
Comment (ashb) [inline airflow/models/taskinstance.py:1829]: I'm not sure how common this case is, and if it's situational/only in our tests, but there was a case from tests when scheduling the same task on 10 dag runs -- it doesn't seem expensive to check this (as soon as the first different task id is found it'll stop) -- so excluding "degenerate" case where 999 tasks have the same task_id and task 1000 has a different one this'll be fine. And even then this is all in memory so should be "fast enough".

WDYT?

Comment (kaxil) [reply airflow/models/taskinstance.py:1829]: A case where there are tasks with same task_id inside the same dag should fail in Master / 2.0

Comment (ashb) [reply airflow/models/taskinstance.py:1829]: Same task_id, but different execution date.

Comment (kaxil) [reply airflow/models/taskinstance.py:1829]: aah yes - it should still be faster than the previous code, so I am fine with this

**Review by kaxil — APPROVED** [review]

**Review by potiuk — COMMENTED** [review]
Nice optimisations overall :)

**Review by potiuk — APPROVED** [review]

**Pull request merged**